### PR TITLE
SOC-903 Facebook Connect

### PR DESCRIPTION
**Pull request by bkoval**
https://wikia-inc.atlassian.net/browse/SOC-903

Facebook Connect handling for existing Wikia accounts.

Includes tracking of all facebook-related events and SOC-1062.

@rafalkalinski @kvas-damian

### CODE REVIEW

File: front/scripts/auth/facebook/FacebookConnect.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+			data = this.getHeliosFacebookConnectData(),
+			url = this.getHeliosFacebookConnectUrl(loginResponse.user_id);
+
+		facebookConnectXhr.onload = (e: Event) => {
+			var status: number = (<XMLHttpRequest> e.target).status;
+
+			if (status === HttpCodes.OK) {
+				window.location.href = this.redirect;
+			} else if (status === HttpCodes.BAD_REQUEST) {
+				//ToDo show the "unable to connect" error
```
Comment (kvas-damian) [inline front/scripts/auth/facebook/FacebookConnect.ts]: Please resolve TODOs

File: front/scripts/auth/facebook/FacebookConnect.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		if (window.location.search) {
+			var params: Object = this.urlHelper.urlDecode(window.location.search.substr(1));
+			this.redirect = params['redirect'];
+		}
+		this.redirect = this.redirect || '/';
+	}
+
+	public init (): void {
+		window.FB.getLoginStatus(function (facebookResponse: FacebookResponse): void {
+			var status = facebookResponse.status;
```
Comment (kvas-damian) [inline front/scripts/auth/facebook/FacebookConnect.ts]: Missing typing for ```status```

File: front/scripts/auth/facebook/FacebookConnect.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	private getHeliosFacebookConnectUrl(userId: string): string {
+		return this.form.getAttribute('data-heliosFacebookConnectURL')
+			+ userId + '/facebook_app_id/' + M.prop('facebookAppId');
+	}
+
+	public onLoginSuccess (loginResponse: LoginResponse): void {
+		var facebookConnectXhr = new XMLHttpRequest(),
+			data = this.getHeliosFacebookConnectData(),
+			url = this.getHeliosFacebookConnectUrl(loginResponse.user_id);
```
Comment (kvas-damian) [inline front/scripts/auth/facebook/FacebookConnect.ts]: missing typings

File: front/scripts/auth/facebook/FacebookConnect.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+class FacebookConnect {
+
+	form: HTMLFormElement;
+	redirect: string;
+	urlHelper: UrlHelper;
+	login: Login;
+
+	constructor (form: HTMLFormElement) {
+		new FacebookSDK(this.init.bind(this));
+		this.login = new Login(form);
```
Comment (kvas-damian) [inline front/scripts/auth/facebook/FacebookConnect.ts]: Did you consider inheritance here?

File: front/scripts/auth/facebook/FacebookConnect.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		facebookConnectXhr.onerror = (e: Event) => {
+			//ToDo show the "unable to connect" error
+		};
+
+		facebookConnectXhr.open('PUT', url, true);
+		facebookConnectXhr.withCredentials = true;
+		facebookConnectXhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
+		facebookConnectXhr.send(this.urlHelper.urlEncode(data));
+	}
+}
```
Comment (kvas-damian) [inline front/scripts/auth/facebook/FacebookConnect.ts]: Don't we need tracking for FacebookConnect?

**Comment by bkoval** [comment]
+ Add tracking

File: server/views/signin-fb.hbs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<section class="auth-content">
+	<form id="loginForm" action="{{heliosLoginURL}}" method="POST" data-heliosFacebookConnectURL="{{heliosFacebookConnectURL}}">
```
Comment (kvas-damian) [inline server/views/signin-fb.hbs]: Please consider using ```signin.hbs``` instead of creating new one?

File: server/views/signin-fb.hbs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+			<input type="text" name="username" id="loginUsername" required />
+			<label class="floating-label" for="loginUsername">{{i18n "auth:common.username-label"}}</label>
+		</div>
+		<div class="input-container">
+			<input class="sign-in-password" type="password" name="password" id="loginPassword" autocomplete="off" required />
+			<label class="floating-label" for="loginPassword">{{i18n "auth:common.password-label"}}</label>
+			<span class="input-icon password-toggler"></span>
+			<a class="forgotten-password" href="{{{forgotPasswordHref}}}" tabindex="-1">{{i18n "auth:signin.forgot-password"}}</a>
+		</div>
+		<button id="loginSubmit" type="submit" disabled><span>{{i18n 'auth:signin.submit-text'}}</span></button>
```
Comment (kvas-damian) [inline server/views/signin-fb.hbs]: Don't we need to update button text?

File: front/scripts/auth/facebook/FacebookConnect.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
+interface HeliosFacebookConnectData {
+	fb_access_token: string;
+}
+
+class FacebookConnect {
+
```
Comment (rafalkalinski) [inline front/scripts/auth/facebook/FacebookConnect.ts]: nowhere else is that first empty line before properties

File: front/scripts/auth/facebook/FacebookConnect.ts (hunk 1 above)
Comment (bkoval) [inline front/scripts/auth/facebook/FacebookConnect.ts]: They will be resolved in SOC-478 (handling validation errors)

File: front/scripts/auth/main.ts [hunk 9]
```diff
@@ -28,7 +28,11 @@ window.document.addEventListener('DOMContentLoaded', function ():void {
 		new SubmitValidator(formElement).watch();
 
 		if (formElement.id === 'loginForm') {
-			new Login(formElement).watch();
+			if (document.body.className.indexOf('fb-connect-page')) {
```
Comment (rafalkalinski) [inline front/scripts/auth/main.ts]: Can't we use some new formElement.id for the facebook connect page, and check only that property?

File: server/views/_layouts/auth.hbs [hunk 10]
```diff
@@ -44,6 +44,9 @@
 		<script>
 			var translations = {{{json (translations language ns='auth-front')}}},
 					language = {{{json language}}};
+			{{#if facebookAppId}}
+				var facebookAppId = {{facebookAppId}};
```
Comment (kvas-damian) [inline server/views/_layouts/auth.hbs]: Please consider using ```pageParams```

File: front/scripts/auth/tracking.ts [hunk 11]
```diff
@@ -2,6 +2,12 @@
 /// <reference path='../mercury/utils/track.ts' />
 
 (function () {
+	var gaCategory = 'user-login-mobile';
```
Comment (kvas-damian) [inline front/scripts/auth/tracking.ts]: Did you consider creating tracker object, that has ```gaCategory``` as a parameter?

**Comment by rafalkalinski** [comment]
LGTM

**Comment by kvas-damian** [comment]
LGTM

**Pull request closed without merging**